crates/client/gateway/src/client/builder.rs

```diff
@@ -48,4 +76,170 @@ impl FeederClient {
                 .expect("Failed to parse Starknet Alpha Sepolia feeder gateway url. This should not fail in prod."),
         )
     }
+
+    pub async fn request_with_pause_handling<F, R>(&self, f: F) -> Result<R, Box<dyn Error + Send + Sync>>
+    where
+        F: FnOnce() -> R,
+    {
+        let pause_until = self.pause_until.clone();
+
+        // Check if a pause is active
+        let pause_duration = {
+            let lock = pause_until.read().await;
+            if let Some(pause_instant) = *lock {
+                let now = Instant::now();
+                if pause_instant > now {
+                    Some(pause_instant - now)
+                } else {
+                    *pause_until.write().await = None;
+                    None
+                }
+            } else {
+                None
+            }
+        };
+
+        // Wait for the pause duration if needed
+        if let Some(duration) = pause_duration {
+            tokio::time::sleep(duration).await;
+        }
+
+        Ok(f())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RetryPolicy {
+    max_retries: usize,
+    backoff: Duration,
+    pause_until: Arc<RwLock<Option<Instant>>>,
+}
+
+impl RetryPolicy {
+    pub fn new(max_retries: usize, backoff: Duration, pause_until: Arc<RwLock<Option<Instant>>>) -> Self {
+        RetryPolicy { max_retries, backoff, pause_until }
+    }
+}
+
+impl Policy<Request<String>, Response<Incoming>, Box<dyn Error + Send + Sync>> for RetryPolicy {
```

Comment:
Consider updating the code to make it more evident this is a retry policy (there also exists `tower::hedge::Policy`)

```suggestion
impl retry::Policy<Request<String>, Response<Incoming>, Box<dyn Error + Send + Sync>> for RetryPolicy {
```